airflow/jobs.py

```diff
@@ -1412,20 +1413,48 @@ def _process_dags(self, dagbag, dags, tis_out):
 
         models.DagStat.update([d.dag_id for d in dags])
 
-    def _process_executor_events(self):
+    @provide_session
+    def _process_executor_events(self, simple_dag_bag, session=None):
         """
         Respond to executor events.
-
-        :param executor: the executor that's running the task instances
-        :type executor: BaseExecutor
-        :return: None
         """
-        for key, executor_state in list(self.executor.get_event_buffer().items()):
+        # TODO: this shares quite a lot of code with _manage_executor_state
+
+        TI = models.TaskInstance
+        for key, state in list(self.executor.get_event_buffer(simple_dag_bag.dag_ids)
+                                   .items()):
             dag_id, task_id, execution_date = key
             self.log.info(
                 "Executor reports %s.%s execution_date=%s as %s",
-                dag_id, task_id, execution_date, executor_state
+                dag_id, task_id, execution_date, state
             )
+            if state == State.FAILED or state == State.SUCCESS:
+                qry = session.query(TI).filter(TI.dag_id == dag_id,
+                                               TI.task_id == task_id,
+                                               TI.execution_date == execution_date)
+                ti = qry.first()
+                if not ti:
+                    self.log.warning("TaskInstance %s went missing from the database", ti)
+                    continue
+
+                if ti.state == State.RUNNING or ti.state == State.QUEUED:
+                    msg = ("Executor reports task instance {} finished ({}) "
```

Comment:
Probably shouldn't handle it in the running case in case this failure does not correspond to this try for some reason. The state should get unset from running anyway because of heartbeats.

Reply:
It is tied to the try. The executor reports on the task_instance which does have the try associated with it. Please note that this exactly the same code as we use for backfills. Basically you want to fail also when running, because the worker is returning and does not manage the task anymore if it is still in RUNNING.

Reply:
In cases where it fails, then is run manually again, then the executor inspects, we will then handle_failure which is probably not ideal. If anything, in the running state, the heartbeat logic  should handle the failure

Reply:
BTW @saguziel : I see what you mean. There is a small change that the scheduler has scheduled a new task instance with the same key and that instance is running now.  While I do think that goes outside the scope of this PR (or maybe not?), we could guard against this by not allowing task_instances known to the executor (i.e. queued_tasks, running, and new: event_buffer) to be executed again or check against the start_date (the 'try' im really not enthusiastic about). Checking against the start_date would require more in-depth changes (start_date should be part of the key) while maybe speeding the scheduler up by a little bit (at best a heartbeat)

Reply:
@saguziel (async comm and no threads): 

`
In cases where it fails, then is run manually again, then the executor inspects, we will then handle_failure which is probably not ideal. If anything, in the running state, the heartbeat logic should handle the failure
`

I don't follow. You mean a kind of race condition? You would need to reset the state first. The task will not start when it finds itself to be in RUNNING

Reply:
IMO, if you only handle it in the queued state, and use a lock, it should be okay.

Reply:
re: 2nd comment:
If the TI is in the running state, the task is either running or not running.
If it is running, we should not interrupt it by calling handle failure on the scheduler.
If it is not, the scheduler heartbeat logic will take care of it.

The executor event in the case when the TI is actually running should able to be discarded.

Reply:
Im not sure if I agree. If the monitor (=worker/executor) reports a failure or success why should we discard that? 

I don't have the heartbeat logic entirely in my mind but what happens if the executor reports SUCCESS and the task stays in RUNNING?

Reply:
If the monitor reports a failure, it is subject to arbitrary delays. So the failure was definitely in the past. If a TI is in RUNNING, we know it is running now (or will be reaped soon), and this is a tested path of Airflow, so we're relatively certain it was the only task running at the time.

If a task stays in RUNNING but is not actually running, it will get set to failure after 2* heartbeat.

Reply:
Been following this thread. I'm wondering if this logic to check for stuck queued TIs from timeout should be in the `kill_zombies` method instead of here, then we can let scheduler heartbeat takes care of all the stale tasks, whether they are said to be running or queued.

I'm not super familiar with the scheduler side of things, but given that all `_process_executor_events` did in the past was logging, it seems to suggest that the executor events are processed somewhere else. May be more clear to rename it to `_log_executor_events`.

Reply:
@jgao54 it can’t be in kill_zombies as there is nothing to kill. The executor actually works as a kind of monitor on the OS level. 

Backfills are managing the executor state and we should have done it here as well.